src/core/model_concrete.jl

```diff
@@ -0,0 +1,257 @@
+
+function construct_agent_container(container, A)
+    if container <: Dict
+        return Dict{Int,A}
+    elseif container <: Vector
+        return Vector{A}
+    else
+        throw(ArgumentError("Unrecognised container $container, please specify either Dict or Vector."))
+    end
+end
+
+"""
+    union_types(U::Type)
+Return a tuple of types within a `Union`.
+"""
+union_types(T::Type) = (T,)
+union_types(T::Union) = (union_types(T.a)..., union_types(T.b)...)
+
+"""
+    SingleContainerABM(AgentType [, space]; properties, kwargs...) → model
+
+A concrete version of [`AgentBasedModel`](@ref) that stores all agents in a
+single container. Offers the variants:
+
+- [`StandardABM`](@ref)
+- [`UnkillableABM`](@ref)
+- [`FixedMassABM`](@ref)
+"""
+function SingleContainerABM(
+    ::Type{A},
+    space::S = nothing;
+    container::Type = Dict{Int,A},
+    scheduler::F = Schedulers.fastest,
+    properties::P = nothing,
+    rng::R = Random.default_rng(),
+    warn = true
+) where {A<:AbstractAgent,S<:SpaceType,F,P,R<:AbstractRNG}
+    agent_validator(A, space, warn)
+    C = construct_agent_container(container, A)
+    agents = C()
+    return SingleContainerABM{S,A,C,F,P,R}(agents, space, scheduler, properties, rng, Ref(0))
+end
+
+function SingleContainerABM(agent::AbstractAgent, args...; kwargs...)
+    return SingleContainerABM(typeof(agent), args...; kwargs...)
+end
+
+
+"""
+    StandardABM(AgentType [, space]; properties, kwargs...) → model
+
+
+The most standard concrete implementation of an [`AgentBasedModel`](@ref),
+as well as the default version of the generic [`AgentBasedModel`](@ref) constructor.
+`StandardABM` stores agents in a dictionary mapping unique `Int` IDs to agents.
+See also [`UnkillableABM`](@ref), [`FixedMassABM`](@ref).
+"""
+StandardABM(args...; kwargs...) = SingleContainerABM(args...; container=Dict)
+
+"""
+    UnkillableABM(AgentType [, space]; properties, kwargs...) → model
+
+Similar to [`StandardABM`](@ref), but agents cannot be removed, only added.
+This allows storing agents more efficiently in a standard Julia `Vector` (as opposed to
+the `Dict` used by [`StandardABM`](@ref), yielding faster retrieval and iteration over agents.
+
+It is mandatory that the agent ID is exactly the same as the agent insertion
+order (i.e., the 5th agent added to the model must have ID 5). If not,
+an error will be thrown by [`add_agent!`](@ref).
+"""
+UnkillableABM(args...; kwargs...) = SingleContainerABM(args...; container=Vector)
+
+"""
+    FixedMassABM(agent_vector [, space]; properties, kwargs...) → model
+
+Similar to [`UnkillableABM`](@ref), but agents cannot be removed nor added.
+Hence, all agents in the model must be provided in advance as a vector.
+This allows storing agents into a `SizedVector`, a special vector with statically typed
+size which is the same as the size of the input `agent_vector`.
+This version of agent based model has slightly better iteration and retrieval speed
+than [`UnkillableABM`](@ref).
```

Comment:
```suggestion
This version of agent based model has slightly better iteration and retrieval speed
than [`UnkillableABM`](@ref).
However, there may be an impact on the compilation speed of the model so it is
advisable to try both container types.
```

Reply:
I wasn't sure exactly what to write here but I felt it important to let users know to at least try both and see what compiles/runs faster.